mpas_analysis/ocean/meridional_overturning_circulation.py

```diff
@@ -522,41 +502,40 @@ def write_file(dsMOCTimeSeries):
 #     return (mocDictClimo, mocDictTseries)
 
 
-def _compute_transport(maxEdgesInTransect, transectEdgeGlobalIDs,
-                       transectEdgeMaskSigns, nz, dvEdge, refLayerThickness,
+def _compute_transport(transectEdgeMaskSigns, nz, dvEdge, refLayerThickness,
                        horizontalVel):  # {{{
 
     '''compute mass transport across southern transect of ocean basin'''
 
-    transportZEdge = np.zeros([nz, maxEdgesInTransect])
-    for i in range(maxEdgesInTransect):
-        if transectEdgeGlobalIDs[i] == 0:
-            break
-        # subtract 1 because of python 0-indexing
-        iEdge = transectEdgeGlobalIDs[i] - 1
-        transportZEdge[:, i] = horizontalVel[iEdge, :] * \
-            transectEdgeMaskSigns[iEdge, np.newaxis] * \
-            dvEdge[iEdge, np.newaxis] * \
-            refLayerThickness[np.newaxis, :]
-    transportZ = transportZEdge.sum(axis=1)
+    transect = transectEdgeMaskSigns != 0
+    transportZ = (refLayerThickness*horizontalVel.where(transect)
+                  *transectEdgeMaskSigns.where(transect)
+                  *dvEdge.where(transect)
+                  ).sum('nEdges')
+
     return transportZ  # }}}
 
 
 def _compute_moc(latBins, nz, latCell, regionCellMask, transportZ,
                  velArea):  # {{{
 
     '''compute meridionally integrated MOC streamfunction'''
-
-    mocTop = np.zeros([np.size(latBins), nz+1])
-    mocTop[0, range(1, nz+1)] = transportZ.cumsum()
-    for iLat in range(1, np.size(latBins)):
-        indlat = np.logical_and(np.logical_and(
-                     regionCellMask == 1, latCell >= latBins[iLat-1]),
-                     latCell < latBins[iLat])
-        mocTop[iLat, :] = mocTop[iLat-1, :] + velArea[indlat, :].sum(axis=0)
+    # note computation should be more cleanly facilitated 
+    # via a multidimensional groupby_bins when available
+    # (see https://github.com/pydata/xarray/pull/924)
+
+    mocTop = xr.DataArray(np.zeros([len(nz), len(latBins)]),
+                          coords=[nz, latBins])
+    mocTop[1:, 0] = transportZ.cumsum(axis=0)
+    for iLat in np.arange(1, len(latBins)):
+        mask = np.logical_and(
+                np.logical_and(latCell >= latBins[iLat-1],
+                               latCell < latBins[iLat]),
+                               regionCellMask > 0)
+        mocTop[:, iLat] = velArea.where(mask).sum('nCells')
+    mocTop = mocTop.cumsum('latBins')
```

Comment:
@pwolfram, well actually the problem was that it really is much more costly to compute the MOCs on each time index and then average than it is to average the velocities and then compute the MOCs.  Since we don't need the global MOC at each time index, just the much smaller Atlantic MOC, not computing the global MOC at each time index seems to lead to huge savings.

And, in this case, loading from the file does still seem to be negligible compared to the computation of the MOC at each time index, though obviously averaging 10 years worth of velocity data is more costly than computing the MOC once.